ouroboros-consensus/test-storage/Test/Ouroboros/Storage/Util.hs

```diff
@@ -83,80 +88,106 @@ expectUnexpectedError unexpectetdErrPred = expectDBError $ \case
-                    <> prettyError e
-                    <> ", but SimFS succeeded.\n\n"
-                    <> "Sim FS: " <> show fs' <> "\n\n"
+          (Left e1, Left e2) -> do
+            assertBool ("SimFS & IOFS didn't return the same error:" <>
+                        "\nSimFS returned: " <> prettyError e1 <>
+                        "\nIOFS  returned: " <> prettyError e2)
+                       (e1 `sameError` e2)
+            -- Doesn't matter if is e1 or e2, as they are equal
+            resAssert (Left e1)
+          (Right (x1, _fs'), Right x2) -> do
+            assertBool ("SimFS & IOFS didn't return the same result:" <>
+                        "\nSimFS returned: " <> show x1 <>
+                        "\nIOFS  returned: " <> show x2)
+                       (x1 == x2)
+            -- Doesn't matter if is x1 or x2, as they are equal
+            resAssert (Right x1)
+          (Left e, Right _) ->
+            fail $ "SimFS returned "
+                <> prettyError e
+                <> ", but IO succeeded.\n\n"
+          (Right (_, fs'), Left e) ->
+            fail $ "IO returned "
+                <> prettyError e
+                <> ", but SimFS succeeded.\n\n"
+                <> "Sim FS: " <> show fs' <> "\n\n"
 
 apiEquivalenceFs :: (HasCallStack, Eq a, Show a)
                  => (Either FsError a -> Assertion)
-                 -> (forall m. (MonadMask m, MonadSTM m, HasCallStack, HasFSE m)
-                            => m (Either FsError a))
+                 -> (forall m. (MonadMask m, MonadSTM m)
+                            => HasFS m
+                            -> ErrorHandling FsError m
+                            -> m a
+                    )
                  -> Assertion
-apiEquivalenceFs = apiEquivalence prettyFsError sameFsError
+apiEquivalenceFs = apiEquivalence tryFS prettyFsError sameFsError
 
 apiEquivalenceDB :: (HasCallStack, Eq a, Show a)
                  => (Either ImmutableDBError a -> Assertion)
-                 -> (forall m. (MonadMask m, MonadSTM m, HasCallStack, HasFSE m)
-                            => m (Either ImmutableDBError a))
+                 -> (forall m. (MonadMask m, MonadSTM m)
+                            => HasFS m
+                            -> ErrorHandling ImmutableDBError m
+                            -> m a
+                    )
                  -> Assertion
-apiEquivalenceDB = apiEquivalence prettyImmutableDBError sameImmutableDBError
+apiEquivalenceDB = apiEquivalence tryImmDB prettyImmutableDBError sameImmutableDBError
+
+tryAny :: IO a -> IO (Either SomeException a)
+tryAny = E.try
+
+tryFS :: IO a -> IO (Either FsError a)
+tryFS = E.try
+
+tryImmDB :: IO a -> IO (Either ImmutableDBError a)
+tryImmDB = fmap squash . E.try . E.try
+  where
+    -- TODO: With the redesigned error handling I'm not sure whether it's
+    -- still necessary that ImmutableDBError can wrap FsError; I think we can
+    -- get rid of that. If we do, this function should become
```

Comment:
But what if you want to do it in a *pure* way, e.g., `Either ImmutableDBError a`? You'd still need a way to embed the `FsError` in `ImmutableDBError`, right? However, maybe we never even need to run it in a *pure* way, as we use `IO` for `MonadSTM` anyway.

Reply:
No, that's not needed. You can just run `ExceptT (Either FsError ImmutableError)` and that would be perfectly fine.